modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/LocalFileConfigurationStorage.java

```diff
@@ -56,99 +66,146 @@
-                    .filter(entry -> entry.getKey().startsWith(prefix))
-                    .collect(Collectors.toMap(Entry::getKey, Entry::getValue));
-            return CompletableFuture.completedFuture(map);
-        } finally {
-            lock.readLock().unlock();
-        }
+        return async(() -> {
+            lock.readLock().lock();
+            try {
+                return latest.entrySet()
+                        .stream()
+                        .filter(entry -> entry.getKey().startsWith(prefix))
+                        .collect(toMap(Entry::getKey, Entry::getValue));
+            } finally {
+                lock.readLock().unlock();
+            }
+        });
     }
 
     @Override
     public CompletableFuture<Serializable> readLatest(String key) {
-        lock.readLock().lock();
-        try {
-            checkAndRestoreConfigFile();
-            return CompletableFuture.completedFuture(latest.get(key));
-        } finally {
-            lock.readLock().unlock();
-        }
+        return async(() -> {
+            lock.readLock().lock();
+            try {
+                return latest.get(key);
+            } finally {
+                lock.readLock().unlock();
+            }
+        });
     }
 
     @Override
     public CompletableFuture<Boolean> write(Map<String, ? extends Serializable> newValues, long ver) {
-        lock.writeLock().lock();
-        try {
-            if (ver != lastRevision) {
-                return CompletableFuture.completedFuture(false);
+        return async(() -> {
+            lock.writeLock().lock();
+            try {
+                if (ver != lastRevision) {
+                    return false;
+                }
+
+                mergeAndSave(newValues);
+
+                sendNotificationAsync(new Data(newValues, lastRevision));
+
+                return true;
+            } finally {
+                lock.writeLock().unlock();
             }
-            checkAndRestoreConfigFile();
-            // TODO: https://issues.apache.org/jira/browse/IGNITE-19152
-            //saveValues(newValues);
-            latest.putAll(newValues);
-            lastRevision++;
-            runAsync(() -> lsnrRef.get().onEntriesChanged(new Data(newValues, lastRevision)));
-            return CompletableFuture.completedFuture(true);
-        } finally {
-            lock.writeLock().unlock();
-        }
+        });
     }
 
-    private void runAsync(Runnable runnable) {
-        CompletableFuture<Void> future = CompletableFuture.runAsync(runnable, threadPool);
+    private void mergeAndSave(Map<String, ? extends Serializable> newValues) {
+        updateLatestState(newValues);
+        saveConfigFile();
+        lastRevision++;
+    }
 
-        futureTracker.registerFuture(future);
+    private void updateLatestState(Map<String, ? extends Serializable> newValues) {
+        newValues.forEach((key, value) -> {
+            if (value == null) { // null means that we should remove this entry
```

Comment:
```suggestion
            if (value == null) { // Null means that we should remove this entry.
```

Reply:
Please don't ignore coding conventions :)

Reply:
Sorry, did not mentioned that it is your suggestion (though it is my code).